train_simple_gpt_v9.ipynb

```diff
@@ -0,0 +1,5192 @@
+              "version_minor": 0,
+              "model_id": "ac3fbec080764890bc17febd4506c3e8"
+            }
+          },
+          "metadata": {}
+        },
+        {
+          "output_type": "display_data",
+          "data": {
+            "text/plain": [
+              "Downloading (…)neration_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]"
+            ],
+            "application/vnd.jupyter.widget-view+json": {
+              "version_major": 2,
+              "version_minor": 0,
+              "model_id": "b972b473335948479272f488acc9cec1"
+            }
+          },
+          "metadata": {}
+        }
+      ],
+      "source": [
+        "from transformers import GPT2Tokenizer, GPT2LMHeadModel\n",
+        "import torch as t\n",
+        "\n",
+        "tokenizer = GPT2Tokenizer.from_pretrained(\"gpt2\")\n",
+        "tokenizer.pad_token = tokenizer.eos_token\n",
+        "pretrained_model = GPT2LMHeadModel.from_pretrained(\"gpt2\")"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "t.cuda.get_device_capability()\n",
+        "#Compute capability score of 7.5."
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "7CyxUWyOSiDP",
+        "outputId": "f4a90c94-139b-4a8e-8950-50b911509558"
+      },
+      "id": "7CyxUWyOSiDP",
+      "execution_count": 13,
+      "outputs": [
+        {
+          "output_type": "execute_result",
+          "data": {
+            "text/plain": [
+              "(7, 5)"
+            ]
+          },
+          "metadata": {},
+          "execution_count": 13
+        }
+      ]
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "#!pip install torchtriton --extra-index-url \"https://download.pytorch.org/whl/nightly/cu117\"\n",
+        "# Make sure we're using a NVIDIA GPU\n",
+        "if t.cuda.is_available():\n",
+        "  gpu_info = !nvidia-smi\n",
+        "  gpu_info = '\\n'.join(gpu_info)\n",
+        "  if gpu_info.find(\"failed\") >= 0:\n",
+        "    print(\"Not connected to a GPU, to leverage the best of PyTorch 2.0, you should connect to a GPU.\")\n",
+        "\n",
+        "  # Get GPU name\n",
+        "  gpu_name = !nvidia-smi --query-gpu=gpu_name --format=csv\n",
+        "  gpu_name = gpu_name[1]\n",
+        "  GPU_NAME = gpu_name.replace(\" \", \"_\") # remove underscores for easier saving\n",
+        "  print(f'GPU name: {GPU_NAME}')\n",
+        "\n",
+        "  # Get GPU capability score\n",
+        "  GPU_SCORE = t.cuda.get_device_capability()\n",
+        "  print(f\"GPU capability score: {GPU_SCORE}\")\n",
+        "  if GPU_SCORE >= (8, 0):\n",
+        "    print(f\"GPU score higher than or equal to (8, 0), PyTorch 2.x speedup features available.\")\n",
+        "  else:\n",
+        "    print(f\"GPU score lower than (8, 0), PyTorch 2.x speedup features will be limited (PyTorch 2.x speedups happen most on newer GPUs).\")\n",
```

Comment:
Interesting. I understand GPU score is equivalent to versioning. Do we know what features are present on newer GPUs that Pytorch 2.x takes advantage of?

Reply:
I don't have a great answer to this at the moment. Been referring to documentation and started with this https://colab.research.google.com/github/mrdbourke/pytorch-deep-learning/blob/main/extras/pytorch_2_intro.ipynb which suggests we need better GPUs.